### ISSUE TITLE
Compilation error when matching reference to empty enum

### ISSUE THREAD

**Issue by dhedey** [rust-lang/rust#131452]
### Summary

I tried this code:

```rust
enum A {}

fn f(a: &A) -> usize {
    match a {}
}
```

I expected to see this happen: This compiles successfully.

Instead, this happened: (E0004 compiler error, expandable below)
<details><summary>E0004 Compiler Error</summary>
<p>

```text
error[E0004]: non-exhaustive patterns: type `&A` is non-empty
 --> src/lib.rs:4:11
  |
4 |     match a {}
  |           ^
  |
note: `A` defined here
 --> src/lib.rs:1:6
  |
1 | enum A {}
  |      ^
  = note: the matched value is of type `&A`
  = note: references are always considered inhabited
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern as shown
  |
4 ~     match a {
5 +         _ => todo!(),
6 +     }
  |

For more information about this error, try `rustc --explain E0004`.
error: could not compile `playground` (lib) due to 1 previous error
```

</p>
</details>

Searching for "Rust E0004" links to docs that don't explain why references to uninhabited types need to be matched: https://doc.rust-lang.org/error_codes/E0004.html - you have to search for "references are always considered inhabited" which takes you to [this issue](https://github.com/rust-lang/rust/issues/78123) from 2023 - more on this history in the Background section below.

### Motivating example

This comes up commonly when creating macros which generate enums, e.g. a very simplified example ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=330d4162e876893ae70b6de80fe71983)):

```rust
macro_rules! define_enum {
    (
        $name: ident,
        [
            $(
                $variant: ident => $string_label: expr
            ),* $(,)?
        ]$(,)?
    ) => {
        enum $name {
            $($variant,)*
        }

        impl $name {
            fn label(&self) -> &'static str {
                match self {
                    $(Self::$variant => $string_label,)*
                }
            }
        }
    }
}

// This compiles
define_enum!{
    MyEnum,
    [Foo => "foo", Bar => "bar"],
}

// This fails compilation with E0004
define_enum!{
    MyEmptyEnum,
    [],
}
```

<details><summary>Compiler Error</summary>
<p>

```text
error[E0004]: non-exhaustive patterns: type `&MyEmptyEnum` is non-empty
  --> src/lib.rs:16:23
   |
16 |                   match self {
   |                         ^^^^
...
31 | / define_enum!{
32 | |     MyEmptyEnum,
33 | |     [],
34 | | }
   | |_- in this macro invocation
   |
note: `MyEmptyEnum` defined here
  --> src/lib.rs:32:5
   |
32 |     MyEmptyEnum,
   |     ^^^^^^^^^^^
   = note: the matched value is of type `&MyEmptyEnum`
   = note: references are always considered inhabited
   = note: this error originates in the macro `define_enum` (in Nightly builds, run with -Z macro-backtrace for more info)
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern as shown
   |
16 ~                 match self {
17 +                     _ => todo!(),
18 +                 }
   |

For more information about this error, try `rustc --explain E0004`.
error: could not compile `playground` (lib) due to 1 previous error
```

</p>
</details>

The fact that this doesn't work for empty enums is quite a gotcha, and I've seen this issue arise a few times as an edge case which isn't only caught a few months after.

But why would we ever use such a macro to generate empty enums? Well this can fall out very natural when generating a hierachy of enums, where some inner enums are empty, e.g.

```rust
define_namespaced_enums! {
    Foo => {
        Fruit: [apple, pear],
        Vegetables: [],
    }
    Bar => {
        Fruit: [banana],
        Vegetables: [carrot],
    }
}
``` 

### Workarounds

Various workarounds include:
* Add a blanket `_ => unreachable!("Workaround for empty enums: references to uninhabited types are considered inhabited at present")`
* In some cases where the enum is copy, use `match *self`
* Add a non-obvious workaround which counts the number of branches, and changes the code generated based on whether the enum is empty or not. And e.g.:
    * Using `match *self {}`
    * Using `!` in the body [as suggested by Ralf](https://github.com/rust-lang/unsafe-code-guidelines/issues/413#issuecomment-2401792481) which is non-stable.

### Background

This was previously raised in this issue: https://github.com/rust-lang/rust/issues/78123 but was closed as "expected behaviour" - due to the fact that:
> ... the reason that &Void (where Void is an uninhabited type) is considered inhabited is because unsafe code can construct an &-reference to an uninhabited type (e.g. with ptr::null() or mem::transmute()). I don't know enough to be sure though.

However, when I raised this as a motivating example in https://github.com/rust-lang/unsafe-code-guidelines/issues/413#issuecomment-2399672350, @RalfJung suggested I raise a new rustc issue for this, and that actually the `match` behaviour is independent of the UB semantics decision:

> So please open a new issue (in the rustc repo, not here) about the macro concern. That's a valid point, just off-topic here. Whether we accept that match code is essentially completely independent of what we decide here.

### Meta

`rustc --version --verbose`:
```
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: aarch64-apple-darwin
release: 1.81.0
LLVM version: 18.1.7
```

Also works on nightly.

**Comment by RalfJung** [comment]
For more background (but partially outdated), also see [this blog post](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/) by @nikomatsakis.

Cc @rust-lang/opsem @rust-lang/types @Nadrieril 

My comment referred to the fact that the safety invariant is uncontroversial in your example, and sufficient to ensure that without unsafe code, nothing weird can happen -- in the specific case considered here, where the match is on a reference.

There are related, more subtle cases though. For instance, what if `ptr: *const !`/`*const Void`, and now we do `match *ptr {}`? Such a pointer can be constructed in safe code, and `let _ = *ptr;` is not UB and might even be(come) safe. Similarly, `match *ptr { _ => () }` is not UB. But with an empty match arm now suddenly this is elevated to UB, even though no safety invariant was violated going into the `match`. IMO that's a footgun we want to avoid: the absence of code introduces UB into a program that doesn't ever obviously violate any validity invariant. (Constructing a place of type `!` is fine. It's reading the discriminant that's a problem, but whether and when a `match` reads the discriminant depends on many factors, and in this case it is strictly the absence of a match arm that is the only thing this code did wrong.)

Also for references, we may allow code to temporarily break the `&Enum` safety invariant (by not making it a validity invariant for the discriminant to be valid). Not sure if we should be concerned about that, but it is something to consider.

So my position on this heavily depends on whether the place we are matching on is "safe" or not. A place is "unsafe" if it involves `*` on a raw pointer or a union field. For unsafe places, I think we need to be careful. I think we want to introduce `!` patterns, not accept anything we don't already accept, and lint against existing code like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7a0d6021819636da59c74a2408cff10d) to enforce the use of never patterns any time "empty type" reasoning is done on an unsafe place.

For safe places, I am much less concerned. Maybe we want an opt-in lint that recommends never patterns to make this explicit. Maybe we want the lint to be warn-by-default when there is unsafe code nearby? But that seems hard and we don't have any other lint like this.

There's also the question of irrefutable patterns exploiting empty types, such as `let OK(foo) = place`. There we can't suggest never patterns as that would destroy the entire point of these patterns. So IMO we should not do empty type reasoning here for unsafe places, i.e. [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=82cebd29ffc65eaa34e816deaa23bf30) and [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=48b21b649387b7a474ccc51ad3229399) should keep being rejected. Though arguably by writing `Ok(_)` the programmer *did* explicitly request a discriminant to be read, and `Result<(), !>` simply doesn't have a discriminant for its `Err` variant, so this is less bad in my view than the `match *ptr {}` case.

**Comment by ijackson** [comment]
Can't we make a distinction between unsafe and safe places and auto-scrutinise the discriminant of safe places?

IOW declare that `&Void` is unsound and say that `*Void` is allowed but you can't do the empty match on it.

**Comment by RalfJung** [comment]
Yes, that is basically what I am suggesting. It is a new distinction between two kinds of places that I don't think we already have, so it's not an entirely trivial extension of the status quo.

**Comment by compiler-errors** [comment]
>  It is a new distinction between two kinds of places that I don't think we already have, so it's not an entirely trivial extension of the status quo.

Well, since `min_exhaustive_patterns` was stabilized, we considers some places to be "exhaustive" and some not to be, depending on their expression and their type... so we kinda already do this, at least in the THIR.

**Comment by Nadrieril** [comment]
+1 to most of what Ralf just said. I'm in favor of making the `&` case convenient at the cost of footguns for users of unsafe references (that's full of footguns already anyway).

> `Result<(), !>` simply doesn't have a discriminant for its `Err` variant

As far as I remember we [hadn't yet committed to that](https://github.com/rust-lang/unsafe-code-guidelines/issues/443), at least in more complex cases like `Result<(), (!, T)>`.

> `let OK(foo) = place`. There we can't suggest never patterns

We can suggest `let (Ok(foo) | Err(!)) = place`.

**Comment by stephanemagnenat** [comment]
As a user perspective note, this pattern happen when one wants to implement the "[Trees that Grow](https://www.jucs.org/jucs_23_1/trees_that_grow/jucs_23_01_0042_0062_najd.pdf)" pattern in Rust:
```rust
enum Never {}

trait Phase {
    type PhaseData;
}

struct PhaseWithInhabited;

impl Phase for PhaseWithInhabited {
    type PhaseData = Never;
}

enum Ast<P: Phase> {
    Identifier(&'static str),
    MaybeSomething(P::PhaseData)
}

fn main() {
    let ast = Ast::<PhaseWithInhabited>::Identifier("hi");
    let s = match &ast {
        Ast::Identifier(s) => s
    };
}
```

Tada: `error[E0004]: non-exhaustive patterns: `&Ast::MaybeSomething(_)` not covered`

Note, as a user, I'll be happy if the compiler would fill the inhabited pattern with `panic`, in case someone does something strange with unsafe code. What is painful is to have to have to manually book-keep inhabited match arms and face the surprising change of behaviour that happens when (`ast` in this case) is changed from passed by value to passed by reference.

**Comment by RalfJung** [comment]
With https://github.com/rust-lang/rfcs/pull/3719, you could write this as
```rust
    let s = match &ast {
        Ast::Identifier(s) => s
        ! // explicitly mark other cases as impossible
    };
```

**Comment by stephanemagnenat** [comment]
That would of course be better than the status quo! It would still be a bit unexpected for new users that the match by value or by reference have different semantics regarding inhabited cases, but if there is a good reason, I guess everyone can live with that!

**Comment by RalfJung** [comment]
For matching on references I'd say this is deep in the undecided territory.  For matching on `*ptr` where `ptr` is a raw pointer, IMO it'd be a mistake to do reasoning like this completely implicitly.

**Comment by stephanemagnenat** [comment]
Yes, my concern is only about references, as they are very common in safe code (as in my AST example). For raw pointers and unsafe code, I have no opinion, as people writing unsafe code are (hopefully) typically way more advanced into Rust than people writing safe code, so they could certainly bear the increased cognitive load of a more complex mental model.

**Comment by dhedey** [comment]
The [explicit never patterns RFC](https://github.com/rust-lang/rfcs/pull/3719/files) is exciting 👍

I just wanted to quickly explore (below) how it might relate to satisfying the discussed safe-rust use cases benefit from stabilisation of [explicit never patterns](https://github.com/rust-lang/rfcs/pull/3719/files), without an [auto-never](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/) sugaring (or something equivalent such as [exhaustive-patterns](https://github.com/rust-lang/rust/issues/51085)).

To summarize in the introduction, never patterns are mostly intended for use in unsafe code, but they won't by themselves allow "safe land" to feel seamless without something like the [auto-never rules](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/) or the `exhaustive_patterns` feature.

### Latest example

In this latest example ([playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b226e9d5a2eb0799d1a34f6f39e9d111)), according to the incomplete `never_patterns` feature, `!` isn't enough, you have to explicitly mention the variants which contain the `!`:

```rust
    let _s = match &ast {
        Ast::Identifier(s) => s,
        Ast::MaybeSomething(!),
    };
```
Although I think there is discussion on the current RFC about possibly loosening that? I've [posted a comment on the RFC](https://github.com/rust-lang/rfcs/pull/3719/files#r1837342309) clarifying.

### Macro-generated nested enum example

Looking at the nested enum example from the `Motivating example` in the original post:

* Under the existing incomplete [never_patterns](https://github.com/rust-lang/rust/issues/118155) feature:
  * This can't currently be solved just by sticking a raw `!` at the root of the `match`.
  * I thought that perhaps you could use the never pattern at each layer of the nesting - e.g. by simply inserting a `!` pattern in each of the sub-matches ([playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9b83821130e6690d29f14c8809595c54)) - we get the compile error `a never pattern must be used on an uninhabited type`.
  * This makes sense, from the point of view of the never pattern making it more explicit what accesses the compiler will make - but it's still a problem for macro code, as it still needs to output different code, branched on if there are 0 variants, or fall back to a workaround like "always output a `_ => unreachable!()` pattern".
* I'm unsure how this changes (if at all) with the RFC

### Side node

I wanted to share my own (ineloquent - and hopefully not too inaccurate) explanation of some of the complexities on display here [inspired by this great one by Nadrieril](https://github.com/rust-lang/rust/issues/51085#issuecomment-743807861) to attempt to help future me and other newcomers come up to speed on the complexities of this:

In "*safe land*" dealing with references, the verbosity of never patterns feel unergonomic because references to uninhabited variants cannot be constructed in safe code (they don't satisfy [the "safety" invariant](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html) of a type which can be assumed in safe code). So, in safe code, the match patterns feel like boilerplate to satisfy the compiler. The code compiling with a lack of patterns would communicate the same meaning as explicit never patterns more succinctly.

But this _isn't true_ in "*unsafe land*". In unsafe land, we care about whether [something is valid](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html) (i.e. whether all code including unsafe code has to uphold this invariant - and therefore that the compiler can assume it holds and optimize using this knowledge).

In most cases, constructing an invalid value is UB, although as an extra complexity, [raw pointers and inside union fields](https://github.com/rust-lang/rfcs/pull/3719#discussion_r1818423527) use an [access-based validity model](https://www.ralfj.de/blog/2018/08/07/stacked-borrows.html) where we only get UB if an invalid thing is accessed. For these types, if you are matching on pointers / references to partially valid values, whether we get UB is dependent on exactly how the compiler processes the match statement, and whether it "accesses" certain pointers/references and asserts the validity of them (or not). There is a [separate thread on whether references to uninhabited types can always be assumed invalid](https://github.com/rust-lang/unsafe-code-guidelines/issues/413), but it's assumed pointers to them can be valid.

So there is clarity benefit to making the resolution of the match statement more explicit, so the accesses (and so the performed validity assertions) are clear from the structure (e.g. through [never patterns](https://github.com/rust-lang/rfcs/pull/3719))

**Comment by RalfJung** [comment]
That sounds pretty good. :)

Personally I am undecided at how far we want to go with "automatic never patterns" in safe code. But I can see how having to write `!` for a match arm that safe code could never possibly hit can feel like boilerplate.

**Comment by stephanemagnenat** [comment]
> So there is clarity benefit to making the resolution of the match statement more explicit

In the case you mentioned, and in interaction with unsafe code, yes. But when using never types to make some enum variants vanish, which is the whole idea of the "Tree that Grow" mentioned above (in that pattern one could have several compiler phases and more than one branch that vanish in each phase), that is a maintenance hindrance. So I guess there are tradeoffs between various use cases.

**Comment by dhedey** [comment]
> Personally I am undecided at how far we want to go with "automatic never patterns" in safe code. But I can see how having to write ! for a match arm that safe code could never possibly hit can feel like boilerplate.

In my mind, there's a compelling argument in terms of user impact - I imagine ~2 orders of magnitude more people are writing matches or destructurings with partially uninhabited types / references (e.g. infallible results) than matching on pointers to partially valid objects.

I generally like explicitness/awkwardness where it prompts the user with a necessary choice they need to handle / consider; but in safe-land code, never patterns just give the awkwardness, and don't have any pay-off in terms of user learning or guidance.

From a rust project organisation perspective, do you think this is the right place to discuss what auto-never could look like? Or is this best left to a different channel / medium?

**Comment by dhedey** [comment]
Musings I have around a potential [auto-never](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/) feature include:

* What are the rules around when an auto-never sugaring might apply? Is it purely about the type? Or about whether we’re in an `unsafe` block?
* Do we add a lint for circumstances where auto-never will apply, but could conceivably be a footgun? Such a lint could be silenced by explicitly writing out never patterns (e.g. outside of an `unsafe` block but near unsafe code or code involving pointers, where a user might be matching on a reference to an invalid value. Such a reference is likely meets the [validity variant](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html) (see [discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/412); [related detail](https://github.com/rust-lang/unsafe-code-guidelines/issues/346)), but not the [safety variant](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html) outside of “unsafe land” but could exist between unsafe blocks in unsafe land) 
* Could we allow auto-never to be explicitly enabled/disabled (say, with an attribute on the match statement)? How would it handle e.g. the partially filled union fields example (u32, !)?
  * On reflection, I imagine it might be not useful in practice. Perhaps an additional lint of "always require explicit never patterns" could be added instead, which would be more flexible, and could achieve the same aim if a user puts it on a setting which prevents compilation if set.

**Comment by RalfJung** [comment]
> From a rust project organisation perspective, do you think this is the right place to discuss what auto-never could look like? Or is this best left to a different channel / medium?

I think it makes more sense to either make this part of the never pattern RFC, or a dedicated companion RFC, depending on what @Nadrieril prefers.

**Comment by dhedey** [comment]
> > So there is clarity benefit to making the resolution of the match statement more explicit
> 
> In the case you mentioned, and in interaction with unsafe code, yes. But when using never types to make some enum variants vanish, which is the whole idea of the "Tree that Grow" mentioned above (in that pattern one could have several compiler phases and more than one branch that vanish in each phase), that is a maintenance hindrance. So I guess there are tradeoffs between various use cases.

Yes, sorry my post was unclear. The clarity benefit was supposed to just be a benefit in `unsafe` land, I've edited the post to make this clearer.

For what it's worth I agree with you - my personal view is that in safe code, requiring explicit never patterns would result in slightly lower clarity, because the boilerplate is irrelevant, and is therefore a distraction from actual code/logic that reader might need to care about.

**Comment by ijackson** [comment]
> my personal view is that in safe code, requiring explicit never patterns would result in slightly lower clarity, because the boilerplate is irrelevant, and is therefore a distraction from actual code/logic that reader might need to care about.

Quite so.  Additionally, it sets macro authors up for failure.  Every macro-generated `match` statement needs to have a spurious `!` arm, or the macro will fail to compile, sometimes, with an empty enum.

**Comment by Nadrieril** [comment]
I agree with all of this. I would prefer for this to be a separate discussion from never patterns: never patterns are useful on their own, and once we have them we can choose how much to elide them. I think this here issue is a good place to discuss this, with the option of going to the rust-lang [Zulip](https://rust-lang.zulipchat.com/) to discuss finer points.

**Comment by clarfonthey** [comment]
Forgive me for not fully understanding the circumstances, but what's confusing me is this:

> There are related, more subtle cases though. For instance, what if `ptr: *const !`/`*const Void`, and now we do `match *ptr {}`? Such a pointer can be constructed in safe code, and `let _ = *ptr;` is not UB and might even be(come) safe. Similarly, `match *ptr { _ => () }` is not UB. But with an empty match arm now suddenly this is elevated to UB, even though no safety invariant was violated going into the `match`. IMO that's a footgun we want to avoid: the absence of code introduces UB into a program that doesn't ever obviously violate any validity invariant. (Constructing a place of type `!` is fine. It's reading the discriminant that's a problem, but whether and when a `match` reads the discriminant depends on many factors, and in this case it is strictly the absence of a match arm that is the only thing this code did wrong.)

Even inside a `match`, shouldn't dereferencing *any* pointer be unsafe, and that be the source of the UB? To me, the UB here isn't that we're matching on an uninhabited reference, but rather that we've created an uninhabited reference from an invalid pointer.

Like, I get that this may not be how the compiler works in reality (still treating references as pointers, and thus having real pointers to nonexistent types even if there's no way to actually form them) but that just feels like the proper takeaway from this.

I also checked to make sure I'm not being incorrect, and the following code requires `unsafe` still:

```rust
enum Void {}
const VOID: *const Void = core::ptr::dangling();
fn main() { match *VOID {} }
````

**Comment by RalfJung** [comment]
There's no reference being created here, so I can't follow what you are saying, sorry.

**Comment by dhedey** [comment]
> Even inside a match, shouldn't dereferencing any pointer be unsafe, and that be the source of the UB? To me, the UB here isn't that we're matching on an uninhabited reference, but rather that we've created an uninhabited reference from an invalid pointer.

I think there's possibly a bit of confusion in your message between e.g. pointers and references; and the difference between unsafe and UB.

If the deep complexities of these topics are new to you (as they were to me a month or two ago - and I'm still coming up the curve), I'd suggest the following reading:
* https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html - captures the difference between a safety invariant and a validity invariant (broadly - the difference between "these are properties which a developer can always expect to hold in 'safe land' code" and "these are properties that always must hold (on pain of UB)". In particular, types can sit in the "should not exist or be operated on in safe code" space but still appear (and possibly be useful) in unsafe code. When paired with match statements, because how a match expands can add "accesses" to sub-parts of the place in the match expression, resulting in assertions of validity against these sub-parts. Roughly speaking this can turn unsafe types into invalid types, i.e. turn unsafe but possibly habitable values into UB.
* https://www.ralfj.de/blog/2018/08/07/stacked-borrows.html - an introduction to thinking about access-based validity (which is used for pointer accesses and union field accesses)
* https://www.ralfj.de/blog/2024/08/14/places.html - an introduction to places

In terms of how it relates to this thread - a lot of this detail is thankfully irrelevant in safe-land code dealing with references (assuming unsafe-land code maintains its safety invariants!) so an ideal solution would have safe-land code not care about this complexity, and unsafe-land code have intuitive tools to handle this. Roughly speaking, the current direction looks to be:
* Never patterns for unsafe code ([see current RFC](https://github.com/rust-lang/rfcs/pull/3719)).
* Some form/equivalence of auto-never for safe code ([see 2018 blog post](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/)).

The right order of tackling these appears a bit unclear, as any separate discussion ends up becoming a bit tangled up. But I'm sure Nadrieril will update us when they have a plan.

**Comment by clarfonthey** [comment]
I guess that my source of confusion here is still that I was under the impression that coercing a pointer into a reference comes with a few contracts that must be upheld, on pain of UB. Even though `&*ptr` doesn't read the pointer, it requires an unsafe block because it's coercing a pointer to a reference, and because safe code makes particular assumptions about the reference, you must ensure those still hold when creating the reference.

For example, I was under the assumption that simply creating a reference from an unaligned pointer is UB whether you read that reference or not. I figured that creating a reference from an uninhabited type, which literally cannot exist, also fell into this situation. But I guess my confusion here is that there's a case where unsafe code can both create an uninhabited reference *and* would want to avoid treating this immediately as UB, which feels confusing.

----

Reading the 2018 post, it makes a little more sense, but I'm still not entirely convinced. The specific case concerns unions, particularly `MaybeUninit`, which are already unsafe to match in general. Again, the key issue here, to me, is that we're not just dealing with pointers, but actual references, and which come with their own validity constraints.

Like, I'm fine with the idea of creating references not being the source of UB, and instead accessing them. All this is fine. What's particularly bothersome to me is the implication that `match x { ! }` or even `match x { &! }` is not  clearly an access, since it feels clear that it's an assertion that the type is empty which could invoke UB if it's not.

Like, don't get me wrong, I love never patterns as an explanation tool, but I'm still not convinced they should be added as a language syntax, because the positives don't really outweigh the negatives. Like, you already need to consider several non-obvious aspects of types in unsafe code, liki ZSTs for example, and I'm not sure that uninhabitedness assertions are weird enough that they deserve gating reference patterns behind an entirely new, custom syntax for it.

**Comment by RalfJung** [comment]
> For example, I was under the assumption that simply creating a reference from an unaligned pointer is UB whether you read that reference or not. 

That is correct. It is also entirely irrelevant in this issue since the relevant examples do not involve any references.

>  there's a case where unsafe code can both create an uninhabited reference and would want to avoid treating this immediately as UB, which feels confusing.

No, that is not the case.

**Comment by Nadrieril** [comment]
> What's particularly bothersome to me is the implication that match x { ! } or even match x { &! } is not clearly an access, since it feels clear that it's an assertion that the type is empty which could invoke UB if it's not.

I'm confused: the whole point of `!` is that they _are_ an access. That's their only purpose.

> it's clear I flipped between being against and for access-based UB in my explanation

There's nothing to be for or against, at least in the context of this RFC: the kind of "access-based" that we're talking about is a core fact of rust unsafe semantics, it's not up for debate here.

It seems to me that you're not familiar with the specific operational semantics context that's required to understand this RFC. This is not the place to explain this context, would you mind continuing this conversation on Zulip?

**Comment by clarfonthey** [comment]
I have a particular aversion to using Zulip for discussion for a number of reasons, although I will stop discussing this here since you're right that this isn't the place. However, I will add that I think that providing the operational semantics context specifically for this issue is an important piece of justifying why this particular change is blocked on some kind of never patterns, and I don't think that this issue nor the proposed RFC do a good enough job of explaining that.

**Comment by camsteffen** [comment]
I feel like there ought to be a special case specifically for empty match on an empty enum. `match` already has the property that it is able to implicitly look past references (I can match on `&T` and write patterns of type `T`). So there is an intuition for me that Rust should be able to see through the reference in the vacuous case and consider the match to be exhaustive. We could still require, conservatively, that more complex types require destructuring to get the inner uninhabited value.
